### PR TITLE
New vehicles

### PR DESCRIPTION
**Pull request by belgien**
- Fixes a bug in carspawner with offset length
- Adds 6 new fully rigged vehicles
- Allows car spawner to spawn different vehicles by random

### CODE REVIEW

File: TrafficSimulator/Assets/Utilities/CarSpawner.cs [hunk 1]
```diff
@@ -15,6 +15,13 @@ public class CarSpawner : MonoBehaviour
     {
         [Header("Connections")]
         [SerializeField] private GameObject _carPrefab;
+        [SerializeField] private GameObject _sportCarPrefab;
+        [SerializeField] private GameObject _sportCar1Prefab;
+        [SerializeField] private GameObject _suvCarPrefab;
+        [SerializeField] private GameObject _suvCar1Prefab;
+        [SerializeField] private GameObject _vanCarPrefab;
+        [SerializeField] private GameObject _vanCar1Prefab;
```
Comment (hannes44) [inline TrafficSimulator/Assets/Utilities/CarSpawner.cs:23]: Should name the prefabs _vanCar1Prefab and _vanCar2Prefab

File: TrafficSimulator/Assets/Utilities/CarSpawner.cs [hunk 2]
```diff
@@ -50,7 +59,9 @@ public class CarSpawner : MonoBehaviour
         
         private void Start()
         {
-            _carLength = _carPrefab.transform.GetChild(1).GetComponent<MeshRenderer>().bounds.size.z;
+            _vehicleTypes = new List<GameObject>(){_carPrefab, _sportCarPrefab, _sportCar1Prefab, _suvCarPrefab, _suvCar1Prefab, _vanCarPrefab, _vanCar1Prefab};
```
Comment (marc7s) [inline TrafficSimulator/Assets/Utilities/CarSpawner.cs:62]:
```suggestion
            _vehicleTypes = new List<GameObject>(){ _carPrefab, _sportCarPrefab, _sportCar1Prefab, _suvCarPrefab, _suvCar1Prefab, _vanCarPrefab, _vanCar1Prefab };
```

File: TrafficSimulator/Assets/AutoDrive/AutoDrive.cs [hunk 3]
```diff
@@ -333,7 +333,9 @@ private void AddSpanNodes(HashSet<LaneNode> spanNodes)
 
             Vector3 direction = _agent.Context.CurrentNode.Position - transform.position;
             float dot = Vector3.Dot(transform.forward, direction.normalized);
-            float distanceToCurrentNode = Vector3.Distance(transform.position, _agent.Context.CurrentNode.Position) * dot;
+
+            float distanceToCurrentNode = (dot < 0) ? -direction.magnitude : direction.magnitude;
+
```
Comment (hannes44) [inline TrafficSimulator/Assets/AutoDrive/AutoDrive.cs:338]: double space

File: TrafficSimulator/Assets/Utilities/CarSpawner.cs [hunk 4]
```diff
@@ -77,6 +88,18 @@ void Update()
...
+        private float GetLongestCarLength(List<GameObject> cars)
+        {
+            float longestCarLength = 0;
+            foreach (GameObject car in cars)
+            {
+                float carLength = car.transform.GetChild(1).GetComponent<MeshRenderer>().bounds.size.z * 1.25f;
+                if (carLength > longestCarLength)
+                    longestCarLength = carLength;
+            }
+            return longestCarLength;
```
Comment (marc7s) [inline TrafficSimulator/Assets/Utilities/CarSpawner.cs:100]:
```suggestion
            return longestCarLength * 1.25f;
```

File: TrafficSimulator/Assets/Utilities/CarSpawner.cs [hunk 5]
```diff
@@ -77,6 +88,18 @@ void Update()
             }
         }
 
+        private float GetLongestCarLength(List<GameObject> cars)
+        {
+            float longestCarLength = 0;
+            foreach (GameObject car in cars)
+            {
+                float carLength = car.transform.GetChild(1).GetComponent<MeshRenderer>().bounds.size.z * 1.25f;
```
Comment (marc7s) [inline TrafficSimulator/Assets/Utilities/CarSpawner.cs:96]:
```suggestion
                float carLength = car.transform.GetChild(1).GetComponent<MeshRenderer>().bounds.size.z;
```

File: TrafficSimulator/Assets/AutoDrive/AutoDrive.cs [hunk 6]
```diff
@@ -333,7 +333,9 @@ private void AddSpanNodes(HashSet<LaneNode> spanNodes)
 
             Vector3 direction = _agent.Context.CurrentNode.Position - transform.position;
             float dot = Vector3.Dot(transform.forward, direction.normalized);
-            float distanceToCurrentNode = Vector3.Distance(transform.position, _agent.Context.CurrentNode.Position) * dot;
+
+            float distanceToCurrentNode = (dot < 0) ? -direction.magnitude : direction.magnitude;
```
Comment (marc7s) [inline TrafficSimulator/Assets/AutoDrive/AutoDrive.cs:337]:
```suggestion
            float distanceToCurrentNode = Vector3.Distance(transform.position, _agent.Context.CurrentNode.Position) * -Mathf.Sign(dot);
```

File: TrafficSimulator/Assets/AutoDrive/AutoDrive.cs [hunk 7]
```diff
@@ -357,7 +359,7 @@ private void AddSpanNodes(HashSet<LaneNode> spanNodes)
             if(!(node.TrafficLight != null && node.TrafficLight.CurrentState == TrafficLightState.Red && node.Intersection?.ID != _agent.Context.PrevIntersection?.ID))
             {
                 node = _agent.Next(_agent.Context.CurrentNode);
-                while (node != null && nodeDistance <= distanceToCurrentNode + _vehicleLength / 2 + VehicleOccupancyOffset + forwardOccupancyOffset)
+                while (node != null && nodeDistance <= -distanceToCurrentNode + _vehicleLength / 2 + VehicleOccupancyOffset + forwardOccupancyOffset)
```
Comment (marc7s) [inline TrafficSimulator/Assets/AutoDrive/AutoDrive.cs:362]:
```suggestion
                while (node != null && nodeDistance <= distanceToCurrentNode + _vehicleLength / 2 + VehicleOccupancyOffset + forwardOccupancyOffset)
```

**Review by marc7s — APPROVED** [review]

**Pull request merged**